Add a skill for benchmarking (and other improvements) (#18)
microsoft/a11y-llm-eval: config/skills/a11y-wizard/components/modal-dialog.md

# Modal dialog

A blocking overlay for focused tasks or destructive confirmations (e.g., "Delete this image?").

## Principles

- A modal dialog has an accessible name and role exposing it as a dialog.
- Opening the dialog moves focus into it (usually the first focusable control, or the primary action for confirmations).
- Background content is inert to AT, keyboard, and pointer while the dialog is open.

> - Focus is trapped inside the dialog until it closes.

Reviewer: ```suggestion
- Keyboard focus can only reach the focusable content of the dialog until it closes.
```

> - Closing the dialog restores focus to the element that opened it.

Reviewer: ```suggestion
- Closing the dialog restores focus to the element that opened it, or another logical location if the element that invoked the dialog no longer exists in the DOM.
```

- `Escape` closes the dialog (if closable); clicking outside MAY close it (depending on destructiveness).
- Dialog content is scrollable if it exceeds the viewport; the background does not scroll.

## Web implementation

### Preferred: native `<dialog>`

- Use `<dialog>` with `.showModal()`. The browser handles focus trap, background inertness, Escape, and the backdrop.

> - Give the dialog an accessible name via `aria-labelledby` referencing the heading inside.

Reviewer: eh.....

what about dialogs that don't have headings?   e.g., a basic "error: you dun messed up. try again." with an "OK" button?

there is a very slow moving effort to have dialogs receive a name from the first nested heading, automatically.  this advice will eventually become outdated - and could even contradict advice on how to purposefully change/negate an unnecessary accName for a dialog.

Reviewer: maybe if this was instead caveated to be about only if a dialog has a heading that would serve as its name?   or indicate this is only needed when a dialog contains a lot of content and a general purpose / name is needed?

### Minimal pattern (native `<dialog>`)

> ```html

Reviewer: should there also be an example of using the `command` and `commandfor` attributes instead of javascript?

```
<button type=button command=show-modal commandfor=d>
  Edit settings
</button>

<dialog id=d>
  <h2>Edit settings</h2>
   ...
   <button command=close commandfor=d>Save</button>
</dialog>
```

<button type="button" id="delete-btn">Delete image</button>

<dialog id="confirm-dialog" aria-labelledby="confirm-heading">
  <h2 id="confirm-heading">Delete this image?</h2>
  <p>This action can't be undone.</p>

  <button type="button" value="cancel" data-close>Cancel</button>
  <button type="button" value="delete" data-confirm>Delete</button>
</dialog>

<script>
  const dialog = document.getElementById('confirm-dialog');
  const trigger = document.getElementById('delete-btn');

  trigger.addEventListener('click', () => dialog.showModal());
  dialog.querySelector('[data-close]').addEventListener('click', () => dialog.close('cancel'));
  dialog.querySelector('[data-confirm]').addEventListener('click', () => dialog.close('delete'));

>   dialog.addEventListener('close', () => trigger.focus());

Reviewer: if the trigger still exists in the DOM, the broswer will automatically return focus.  this shouldn't be needed unless it's updated to an if statement for "if trigger doesn't exist, then focus x element instead"

</script>
```

### Fallback: ARIA dialog

When `<dialog>` isn't an option:

- Container: `role="dialog"` or `role="alertdialog"` (the latter for short destructive confirmations), `aria-modal="true"`, `aria-labelledby` referencing the title, optional `aria-describedby` for the body.

> - Use `inert` on all siblings of the dialog while it's open to make the rest of the page non-interactive.

Reviewer: this should be clearer on the requirements.  

e.g., all siblings of the dialog in the following case wouldnt make everything inert, only the invoking button would be because it's the only sibling:
```
<main>
   <article>....</article>
   <div>
      <button ...>open dialog</button>
      <dialog ...>...</dialog>
   </div>
</main>
```

so another check would be something like "make sure all dialogs are declared as siblings to the element or elements that contain all the content of the web page."
e.g., 
```
<div id=app> <!-- all content here --> </div>
<!-- dialogs here -->

or

<header>...</header>
<main>...</main>
<!-- dialogs go here -->
<footer>...<footer>
```

> - Implement a focus trap: when focus reaches the last focusable element, Tab wraps to the first; Shift+Tab from the first wraps to the last.

Reviewer: technically this shouldn't be necessary if inert was used and the showModal API was used.  if using the alternate declarative command/commandfor method, _then_ this would be needed due to a bug in browsers...but that should hopefully be fixed soon.

> - Handle `Escape` to close.

Reviewer: not all dialogs must be closable by escape key.  

```suggestion
- Handle `Escape` to close - or ensure a keyboard accessible close button is available.
```

> - Restore focus to the trigger on close.

Reviewer: or another logical element if the trigger no longer exists

```html
<div role="alertdialog" aria-modal="true"
     aria-labelledby="confirm-heading" aria-describedby="confirm-body">
  <h2 id="confirm-heading">Delete this image?</h2>
  <p id="confirm-body">This action can't be undone.</p>
  <button type="button">Cancel</button>
  <button type="button">Delete</button>
</div>
```

> ### Review checklist

Reviewer: not going to re-comment on checklist items, but stuff here would be changed based on prior comments.

- Dialog has a programmatic role (`<dialog>` or `role="dialog"` / `role="alertdialog"`) and an accessible name.
- Opening moves focus into the dialog.
- While open, background is inert (native `<dialog>.showModal()` or `inert` on siblings).
- Focus is trapped inside the dialog.
- `Escape` closes the dialog.
- Closing restores focus to the trigger.
- Dialog is reachable only when open; hidden from AT / focus order when closed.
- Content can scroll within the dialog if it overflows; the background doesn't scroll.
- Destructive actions use `role="alertdialog"` or a clearly-named dialog; the default/initial focus does not land on the destructive action.

### Pitfalls

> - Opening a dialog without moving focus — keyboard users are stranded on the trigger.

Reviewer: ```suggestion
- Opening a dialog without moving focus — keyboard users are stranded on the trigger. Screen reader users will receive no confirmation of action for the changed state of the page or dialog.
```

> - Not restoring focus on close — keyboard users lose their place.

Reviewer: ```suggestion
- Not restoring focus on close — keyboard users may lose their place, screen reader users will commonly be returned to the top of the web page.
```

> - Forgetting `inert` on background content in the ARIA fallback — Tab escapes the dialog.

Reviewer: ```suggestion
- Forgetting `inert` on background content in the ARIA fallback — Keyboard tabbing can reach focusable elements outside of the modal dialog.
```

> - Using `role="dialog"` without `aria-modal="true"` and then trapping focus anyway — mismatched semantics.

Reviewer: mmm.... this is maybe going a bit far.  non-modal dialogs can also trap keyboard focus and expect someone to navigate around using ctrl + f6.    i'd take this out.
